src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,1305 @@
+            open_set;
+    const tripoint_abs_ms veh_pos = abs_vehicle_pos( driven_veh );
+    const node_address start = data.nav_to_map.inverse().transform(
+                                   veh_pos.raw().xy(), to_orientation( driven_veh.face.dir() ) );
+    known_nodes.emplace( start, make_start_node( start, driven_veh ) );
+    open_set.push( scored_address{ start, 0 } );
+    int search_count = 0;
+    std::vector<std::pair<node_address, navigation_node>> next_nodes;
+    while( !open_set.empty() ) {
+        const node_address cur_addr = open_set.top().addr;
+        open_set.pop();
+        search_count++;
+        const navigation_node &cur_node = known_nodes[cur_addr];
+        if( cur_node.is_goal ) {
+            node_address addr = cur_addr;
+            while( !( addr == start ) ) {
+                const navigation_node &node = known_nodes.at( addr );
+                const node_address &prev = node.prev;
+                const tripoint_abs_ms prev_loc( data.nav_to_map.transform( prev.get_point(),
+                                                data.current_omt.z() ) );
+                ret.emplace_back( navigation_step{
+                    prev_loc,
+                    data.nav_to_map.transform( addr.facing_dir ),
+                    node.target_speed_tps
+                } );
+                addr = prev;
+            }
+            return ret;
+        }
+        next_nodes.clear();
+        compute_next_nodes( cur_addr, cur_node, speed_tps, next_nodes );
+        for( const auto &next : next_nodes ) {
+            const node_address &next_addr = next.first;
+            const navigation_node &next_node = next.second;
+            auto iter = known_nodes.find( next_addr );
+            if( iter != known_nodes.end() ) {
+                const navigation_node &other_node = iter->second;
+                if( next_node.cost < other_node.cost ) {
+                    const bool same_dir = cur_addr.facing_dir == next_addr.facing_dir;
+                    const bool dir_multiple_45 = static_cast<int>( next_addr.facing_dir ) % 3 == 0;
+                    if( ( !same_dir && next_node.tileray_steps == 1 ) || dir_multiple_45 ) {
+                        iter->second = next_node;
+                    }
+                }
+            } else if( known_nodes.size() < max_search_count ) {
+                known_nodes[next_addr] = next_node;
+                open_set.push( compute_node_score( next_addr, next_node ) );
+            }
+        }
+    }
+    return cata::nullopt;
+}
+
+vehicle::autodrive_controller::autodrive_controller( const vehicle &driven_veh,
+        const Character &driver ) : driven_veh( driven_veh ), driver( driver )
+{
+    data.clear();
+}
+
+void vehicle::autodrive_controller::check_safe_speed()
+{
+    int safe_speed_tps = driven_veh.safe_velocity() / VMIPH_PER_TPS;
+    if( data.max_speed_tps > safe_speed_tps ) {
+        data.max_speed_tps = safe_speed_tps;
+    }
```

Comment:
Down only? OK currently. Will this get reset each time the autodrive has to replan? I'm guessing so given reduce_speed's actions and when it's called.

Reply:
Right now check_safe_speed() exists only to cover the "amphibious vehicle entering water" use case you brought up earlier.
If/when we discover that the vehicle is driving above the safe speed beyond which the engine starts taking damage we immediately reduce speed.

I added a comment.